internal/runner/operation.go

```diff
@@ -248,6 +260,7 @@ func (r *Runner) executeJob(
 		core.WithDataDir(projDir),
 		core.WithLabels(job.Labels),
 		core.WithVariables(inputVars),
+		core.WithVariableRefs(jobVars),
```

Comment:
If you remove variable refs from operations as I suggested, then you can nix this whole code path plus everything in `core`.